R/utils.R

```diff
@@ -171,6 +172,122 @@ list_init_empty <- function(x,
   list(x = x, sizes = sizes, null = empty_null, typed = empty_typed)
 }
 
+#' Replace `NULL` list elements
+#'
+#' @param x A list, but not a list-of.
+#'
+#' @param sizes An integer vector of sizes of the `x` elements.
+#'
+#' @param ptype One of:
+#'
+#'   - `NULL` to fill `NULL` elements with `unspecified()`.
+#'
+#'   - A `ptype` value to fill `NULL` elements with.
+#'
+#' @param size The size of the replacement value to fill with. Commonly either
+#'   `0L` or `1L`.
+#'
+#' @returns
+#' A named list holding `x` with `NULL`s replaced, and `sizes` updated to
+#' have the correct size where `NULL` elements were replaced.
+#'
+#' @noRd
+list_init_null <- function(x, sizes, ..., ptype = NULL, size = 1L) {
+  check_dots_empty0(...)
+
+  if (!vec_is_list(x)) {
+    cli::cli_abort("`x` must be a list.", .internal = TRUE)
+  }
+  if (is_list_of(x)) {
+    cli::cli_abort("`x` can't be a list-of. Unclass first and provide `ptype`.", .internal = TRUE)
+  }
+
+  if (vec_any_missing(x)) {
+    null <- vec_detect_missing(x)
+    null <- which(null)
```

Comment:
Can't you assume that `vec_assign()` will do this?

Reply:
It definitely will. But since I use `null` in two separate assignments below, it seemed slightly more efficient to `which()` it once here